numba/tests/test_linalg.py

```diff
@@ -344,31 +350,84 @@ def sample_vector(self, n, dtype):
         else:
             return (base * 0.5 + 1).astype(dtype)
 
-    def orth_fact_sample_matrix(self, size, dtype, order):
+    def specific_sample_matrix(
+            self, size, dtype, order, rank=None, condition=None):
         """
-        Provides a sample matrix for use in orthogonal factorization tests.
+        Provides a sample matrix with an optionally specified rank or condition
+        number.
 
         size: (rows, columns), the dimensions of the returned matrix.
         dtype: the dtype for the returned matrix.
         order: the memory layout for the returned matrix, 'F' or 'C'.
+        rank: the rank of the matrix, an integer value, defaults to full rank.
+        condition: the condition number of the matrix (defaults to 1.)
+
+        NOTE: Only one of rank or condition may be set.
         """
-        # May be worth just explicitly constructing this system at some point
-        # with a kwarg for condition number?
-        break_at = 10
-        jmp = 0
-        # have a few attempts at shuffling to get the condition number down
-        # else not worry about it
-        mn = size[0] * size[1]
+
+        # default condition
+        d_cond = 1.
+
+        if len(size) != 2:
+            raise ValueError("size must be a length 2 tuple.")
+
+        if order not in ['F', 'C']:
+            raise ValueError("order must be one of 'F' or 'C'.")
+
+        if dtype not in [np.float32, np.float64, np.complex64, np.complex128]:
+            raise ValueError("dtype must be a numpy floating point type.")
+
+        if rank is not None and condition is not None:
+            raise ValueError("Only one of rank or condition can be specified.")
+
+        if condition is None:
+            condition = d_cond
+
+        if condition < 1:
+            raise ValueError("Condition number must be >=1.")
+
         np.random.seed(0)  # repeatable seed
-        while jmp < break_at:
-            v = self.sample_vector(mn, dtype)
-            # shuffle to improve conditioning
-            np.random.shuffle(v)
-            A = np.reshape(v, size)
-            if np.linalg.cond(A) < mn:
-                return np.array(A, order=order, dtype=dtype)
-            jmp += 1
-        return A
+        m, n = size
+
+        if m < 0 or n < 0:
+            raise ValueError("Negative dimensions given for matrix shape.")
+
+        minmn = min(m, n)
+        if rank is None:
+            rv = minmn
+        else:
+            if rank <= 0:
+                raise ValueError("Rank must be greater than zero.")
+            if not isinstance(rank, Integral):
+                raise ValueError("Rank must an integer.")
+            rv = rank
+            if rank > minmn:
+                raise ValueError("Rank given greater than full rank.")
+
+        # Build a sample matrix via combining SVD like inputs
+        # this ought to use MGS
```

Comment:
Modified Gram–Schmidt. Instead of running QR on random data to get an orthnormal matrix, MGS could be used more cheaply, but it doesn't exist in numpy as far as I can see?